packages/ses/src/strict-scope-terminator.js

```diff
@@ -0,0 +1,92 @@
+import {
+  Proxy,
+  String,
+  TypeError,
+  ReferenceError,
+  create,
+  freeze,
+  getOwnPropertyDescriptors,
+  globalThis,
+  immutableObject,
+} from './commons.js';
+import { assert } from './error/assert.js';
+
+const { details: d, quote: q } = assert;
+
+/**
+ * alwaysThrowHandler
+ * This is an object that throws if any property is called. It's used as
+ * a proxy handler which throws on any trap called.
+ * It's made from a proxy with a get trap that throws. It's safe to
+ * create one and share it between all Proxy handlers.
+ */
+export const alwaysThrowHandler = new Proxy(
+  immutableObject,
+  freeze({
+    get(_shadow, prop) {
+      // eslint-disable-next-line @endo/no-polymorphic-call
+      assert.fail(
+        d`Please report unexpected scope handler trap: ${q(String(prop))}`,
+      );
+    },
+  }),
+);
+
+/*
+ * scopeProxyHandlerProperties
+ * scopeTerminatorHandler manages a strictScopeTerminator Proxy which serves as
+ * the final scope boundary that will always return "undefined" in order
+ * to prevent access to "start compartment globals".
+ */
+const scopeProxyHandlerProperties = {
+  get(_shadow, _prop) {
+    return undefined;
+  },
+
+  set(_shadow, prop, _value) {
+    // we should only hit this if the has() hook returned true
+    // matches the v8 ReferenceError message
+    // "Uncaught ReferenceError: xyz is not defined"
+    throw new ReferenceError(`${String(prop)} is not defined`);
+  },
+
+  has(_shadow, prop) {
+    // we must at least return true for all properties on the realm globalThis
+    return prop in globalThis;
+  },
+
+  // note: this is likely a bug of safari
+  // https://bugs.webkit.org/show_bug.cgi?id=195534
+  getPrototypeOf() {
+    return null;
+  },
+
+  // Chip has seen this happen single stepping under the Chrome/v8 debugger.
+  // TODO record how to reliably reproduce, and to test if this fix helps.
+  // TODO report as bug to v8 or Chrome, and record issue link here.
+  getOwnPropertyDescriptor(_target, prop) {
+    // Coerce with `String` in case prop is a symbol.
+    const quotedProp = q(String(prop));
+    // eslint-disable-next-line @endo/no-polymorphic-call
+    console.warn(
+      `getOwnPropertyDescriptor trap on scopeTerminatorHandler for ${quotedProp}`,
+      new TypeError().stack,
+    );
+    return undefined;
+  },
+};
+
+// The scope handler's prototype is a proxy that throws if any trap other
+// than get/set/has are run (like getOwnPropertyDescriptors, apply,
+// getPrototypeOf).
+export const strictScopeTerminatorHandler = freeze(
+  create(
+    alwaysThrowHandler,
+    getOwnPropertyDescriptors(scopeProxyHandlerProperties),
+  ),
+);
+
+export const strictScopeTerminator = new Proxy(
+  immutableObject,
+  strictScopeTerminatorHandler,
+);
```

Comment:
Wondering if it wouldn't make more sense to encapsulate / DRY up the scope proxy making logic a little more.

```suggestion
export const makeScopeTerminatorProxy = customHandlerProperties =>
  new Proxy(
    immutableObject,
    freeze(
      create(
        alwaysThrowHandler,
        getOwnPropertyDescriptors({
          ...scopeProxyHandlerProperties,
          ...customHandlerProperties,
        }),
      ),
    ),
  );

export const strictScopeTerminator = makeScopeTerminatorProxy();
```

Reply:
Here is what it would look like: https://github.com/endojs/endo/commit/dd156de59c4f862d99ed73ba298a4ca902ec4bdd